nrf52-hal-common/src/spim.rs

```diff
@@ -119,79 +185,64 @@ impl<T> Spim<T> where T: SpimExt {
+            return Err(Error::DMABufferInUnaccessableMemory)
         }
-        if rx_buffer.len() > EASY_DMA_SIZE {
-            return Err(Error::RxBufferTooLong);
-        }
-
-        // Pull chip select pin high, which is the inactive state
-        chip_select.set_high();
-
         // Conservative compiler fence to prevent optimizations that do not
         // take in to account actions by DMA. The fence has been placed here,
         // before any DMA action has started
         compiler_fence(SeqCst);
-
+        // set CS inactive
+        cs_n(false);
         // Set up the DMA write
         self.0.txd.ptr.write(|w|
-            // We're giving the register a pointer to the stack. Since we're
-            // waiting for the SPI transaction to end before this stack pointer
-            // becomes invalid, there's nothing wrong here.
-            //
-            // The PTR field is a full 32 bits wide and accepts the full range
-            // of values.
-            unsafe { w.ptr().bits(tx_buffer.as_ptr() as u32) }
+            unsafe { w.ptr().bits(tx_data_ptr) }
         );
         self.0.txd.maxcnt.write(|w|
-            // We're giving it the length of the buffer, so no danger of
-            // accessing invalid memory. We have verified that the length of the
-            // buffer fits in an `u8`, so the cast to the type of maxcnt
-            // is also fine.
-            //
             // Note that that nrf52840 maxcnt is a wider
             // type than a u8, so we use a `_` cast rather than a `u8` cast.
             // The MAXCNT field is thus at least 8 bits wide and accepts the full
             // range of values that fit in a `u8`.
-            unsafe { w.maxcnt().bits(tx_buffer.len() as _) }
+            unsafe { w.maxcnt().bits(tx_len as _ ) }
         );
 
         // Set up the DMA read
         self.0.rxd.ptr.write(|w|
             // This is safe for the same reasons that writing to TXD.PTR is
             // safe. Please refer to the explanation there.
-            unsafe { w.ptr().bits(rx_buffer.as_mut_ptr() as u32) }
+            unsafe { w.ptr().bits(rx_data_ptr ) }
         );
         self.0.rxd.maxcnt.write(|w|
             // This is safe for the same reasons that writing to TXD.MAXCNT is
             // safe. Please refer to the explanation there.
-            unsafe { w.maxcnt().bits(rx_buffer.len() as _) }
+            unsafe { w.maxcnt().bits(rx_len as _) }
         );
 
+        // Set CS active
+        cs_n(true);
         // Start SPI transaction
-        chip_select.set_low();
         self.0.tasks_start.write(|w|
             // `1` is a valid value to write to task registers.
             unsafe { w.bits(1) }
         );
 
+        // Conservative compiler fence to prevent optimizations that do not
+        // take in to account actions by DMA. The fence has been placed here,
+        // after all possible DMA actions have completed
+        compiler_fence(SeqCst);
```

Comment:
What's the reason for this compiler fence? I don't think we need it (but I'm not sure).

Reply:
(bad excuse again )The compiler fences came form the originating code, and I kept them where they landed. Will give it a try and see which could be removed.

Reply:
If in doubt, leaving compiler fences in is a good idea, I think. The worst they can do is prevent some optimizations.

The perfect scenario for me would be, if someone took a detailed look and proclaimed, with great certainty, which compiler fences are required and which aren't. As long as that doesn't happen, I'm happy to merge code that has an additional compiler fence here or there :-)